packages/dds/shared-object-base/src/sharedObjectKernel.ts

```diff
@@ -0,0 +1,469 @@
+ * 2. The app, who uses the SharedObject to read and write data.
+ *
+ * There is some common functionality all shared objects use, provided by {@link SharedObject}.
+ * SharedKernel describes the portion of the behavior required by the runtime which
+ * differs between different kinds of shared objects.
+ *
+ * {@link makeSharedObjectKind} is then used to wrap up the kernel into a full {@link ISharedObject} implementation.
+ * The runtime specific APIs are then type erased into a {@link SharedObjectKind}.
+ * @privateRemarks
+ * Unlike the `SharedObject` class, this interface is internal, and thus can be adjusted more easily.
+ * Therefore this interface is not intended to address all needs, and will likely need small changes as it gets more adoption.
+ *
+ * @internal
+ */
+export interface SharedKernel {
+	/**
+	 * {@inheritDoc SharedObject.summarizeCore}
+	 */
+	summarizeCore(
+		serializer: IFluidSerializer,
+		telemetryContext: ITelemetryContext | undefined,
+		incrementalSummaryContext: IExperimentalIncrementalSummaryContext | undefined,
+	): ISummaryTreeWithStats;
+
+	/**
+	 * {@inheritDoc SharedObjectCore.onDisconnect}
+	 */
+	onDisconnect(): void;
+
+	/**
+	 * {@inheritDoc SharedObjectCore.reSubmitCore}
+	 */
+	reSubmitCore(content: unknown, localOpMetadata: unknown): void;
+
+	/**
+	 * {@inheritDoc SharedObjectCore.applyStashedOp}
+	 */
+	applyStashedOp(content: unknown): void;
+
+	/**
+	 * {@inheritDoc SharedObjectCore.processMessagesCore}
+	 */
+	processMessagesCore(messagesCollection: IRuntimeMessageCollection): void;
+
+	/**
+	 * {@inheritDoc SharedObjectCore.rollback}
+	 */
+	rollback?(content: unknown, localOpMetadata: unknown): void;
+
+	/**
+	 * {@inheritDoc SharedObjectCore.didAttach}
+	 */
+	didAttach?(): void;
+}
+
+/**
+ * SharedObject implementation that delegates to a SharedKernel.
+ * @typeParam TOut - The type of the object exposed to the app.
+ * Once initialized instances of this class forward properties to the `TOut` val;ue provided by the factory.
+ * See {@link mergeAPIs} for more limitations.
+ *
+ * @remarks
+ * The App facing API (TOut) needs to be implemented by this object which also has to implement the runtime facing API (ISharedObject).
+ *
+ * Requiring both of these to be implemented by the same object adds some otherwise unnecessary coupling.
+ * This class is a workaround for that, which takes separate implementations of the two APIs and merges them into one using {@link mergeAPIs}.
+ */
+class SharedObjectFromKernel<
+	TOut extends object,
+	TEvent extends ISharedObjectEvents,
+> extends SharedObject<TEvent> {
+	/**
+	 * Lazy init here so kernel can be constructed in loadCore when loading from existing data.
+	 *
+	 * Explicit initialization to undefined is done so Proxy knows this property is from this class (via `Reflect.has`),
+	 * not from the grafted APIs.
+	 */
+	#lazyData: FactoryOut<TOut> | undefined = undefined;
+
+	readonly #kernelArgs: KernelArgs;
+
+	/**
+	 * @param id - String identifier.
```

Comment:
These docs shouldn't be necessary. Intellisense is smart enough to derive these from the docs on `SharedObject`. 

See https://github.com/microsoft/FluidFramework/pull/24495

Reply:
Wow, thats creepy: it assumes any parameter with a matching name is passed to the constructor. If you don't pass a parameter to the constructor, it will still get its docs, and if either renames the parameter, it will break the docs with no way for a code reviewer to easily notice. Useful, but something to watch out for!

Reply:
Yeah. I think this is exactly why TSDoc's `@inheritDoc` _requires_ you specify what you're inheriting. They took the stance that ambiguity could cause problems, so just make everything explicit. If only intellisense would support it...